src/renderer/atlas/AtlasEngine.r.cpp

```diff
@@ -119,23 +125,103 @@ catch (const wil::ResultException& exception)
 }
 CATCH_RETURN()
 
+[[nodiscard]] bool AtlasEngine::RequiresContinuousRedraw() noexcept
+{
+    return debugGeneralPerformance || _r.requiresContinuousRedraw;
+}
+
+void AtlasEngine::WaitUntilCanRender() noexcept
+{
+    // IDXGISwapChain2::GetFrameLatencyWaitableObject returns an auto-reset event.
+    // Once we've waited on the event, waiting on it again will block until the timeout elapses.
+    // _r.waitForPresentation guards against this.
+    if (!debugGeneralPerformance && std::exchange(_r.waitForPresentation, false))
+    {
+        WaitForSingleObjectEx(_r.frameLatencyWaitableObject.get(), 100, true);
+#ifndef NDEBUG
+        _r.frameLatencyWaitableObjectUsed = true;
+#endif
+    }
+}
+
 #pragma endregion
 
-void AtlasEngine::_setShaderResources() const
+void AtlasEngine::_renderWithCustomShader() const
 {
-    _r.deviceContext->VSSetShader(_r.vertexShader.get(), nullptr, 0);
-    _r.deviceContext->PSSetShader(_r.pixelShader.get(), nullptr, 0);
+    // Render with our main shader just like Present().
+    {
+        // OM: Output Merger
+        _r.deviceContext->OMSetRenderTargets(1, _r.customOffscreenTextureTargetView.addressof(), nullptr);
+        _r.deviceContext->Draw(3, 0);
+    }
+
+    // Update the custom shader's constant buffer.
+    {
+        CustomConstBuffer data;
+        data.time = std::chrono::duration<float>(std::chrono::steady_clock::now() - _r.customShaderStartTime).count();
+        data.scale = _r.pixelPerDIP;
+        data.resolution.x = static_cast<float>(_r.cellCount.x * _r.fontMetrics.cellSize.x);
+        data.resolution.y = static_cast<float>(_r.cellCount.y * _r.fontMetrics.cellSize.y);
+        data.background = colorFromU32<f32x4>(_r.backgroundColor);
+
+#pragma warning(suppress : 26494) // Variable 'mapped' is uninitialized. Always initialize an object (type.5).
+        D3D11_MAPPED_SUBRESOURCE mapped;
+        THROW_IF_FAILED(_r.deviceContext->Map(_r.customShaderConstantBuffer.get(), 0, D3D11_MAP_WRITE_DISCARD, 0, &mapped));

```

Comment:
do we need to re-map the ConstBuffer in every time? Or can we map it once and update it like we do with the other Atlas bits?

Reply:
We need to map/unmap it every time. Here's Nvidia's documentation for it: https://developer.nvidia.com/content/constant-buffers-without-constant-pain-0

Reply:
Ah, thanks!